src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,422 @@
+      refl,
+      end}⟩
+
+@[simp] lemma coe_sub {f g : (modular_form Γ k)} : ⇑(f - g) = f - g := rfl
+@[simp] lemma sub_apply {f g : (modular_form Γ k)} {z : ℍ} : (f - g) z = f z - g z := rfl
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl (λ _ _, coe_add) (λ _, coe_neg) (λ _ _, coe_sub)
+(λ _ _, coe_nsmul) (λ _ _, coe_zsmul)
+
+lemma coe_zero : ((0 : (modular_form Γ k) ) : ℍ → ℂ) = (0 : ℍ → ℂ) := rfl
+
+/--Additive coercieon from `modular_form` to `ℍ → ℂ`. -/
+def coe_hom : (modular_form Γ k) →+ (ℍ → ℂ) :=
+{ to_fun := λ f, f, map_zero' := coe_zero, map_add' := λ _ _, rfl }
+
+lemma coe_hom_injective : function.injective (@coe_hom Γ k) :=
+fun_like.coe_injective
+
+instance : module ℂ (modular_form Γ k) :=
+coe_hom_injective.module ℂ (coe_hom) (λ _ _, rfl)
+
+instance : inhabited (modular_form Γ k) := ⟨0⟩
+
+/--The modular form of weight `k_1+k_2` given by the product of two modular forms of weights
+`k_1` and `k_2`. -/
+def mul (k_1 k_2 : ℤ) (Γ : subgroup SL(2, ℤ)) (f : (modular_form Γ k_1))
+  (g : (modular_form Γ k_2)) : (modular_form Γ (k_1 + k_2)) :=
+{ to_fun := f * g,
+  slash_action_eq' := by {intro A, rw mul_slash_subgroup, congr,
+  apply f.slash_action_eq' A, apply g.slash_action_eq' A,},
+  hol' := f.hol'.mul _ g.hol',
+  bdd_at_infty' := λ A, begin
+    rw [mul_slash_SL2],
+    exact (f.bdd_at_infty' A).mul (g.bdd_at_infty' A),
+  end}
+
+  /-- The constant function is bounded at infinity. -/
+lemma is_bounded_at_im_infty_one : is_bounded_at_im_infty (1 : ℍ → ℂ):=
+@asymptotics.is_O_const_const _ _ ℂ _ _ 1 _ one_ne_zero _
+
+instance : has_one (modular_form Γ 0) :=
+{ one :=
+  { to_fun := (1 : slash_invariant_form Γ 0),
+    slash_action_eq' := (1 : slash_invariant_form Γ 0).2,
+    hol' := (λ (x : ℍ), mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+    bdd_at_infty' := by {intro A,
+      convert is_bounded_at_im_infty_one,
+      apply const_one_form_is_invar A}}}
+
+end modular_form
+
+namespace cusp_forms
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (cusp_form Γ k) :=
+{ add := λ f g,
+  { to_fun := f + g,
+    slash_action_eq' := (f.to_slash_invariant_form + g.to_slash_invariant_form).2,
+    hol' := f.hol'.add _ g.hol',
+    zero_at_infty' := λ A, begin
+      rw slash_action.add_action,
+      exact (zero_at_im_infty_submodule ℂ).add_mem' (f.zero_at_infty' A) (g.zero_at_infty' A),
+      end}}
+
+@[simp] lemma coe_add {f g : (cusp_form Γ k)} : ⇑(f + g) = f + g := rfl
+@[simp] lemma add_apply {f g : (cusp_form Γ k)} {z : ℍ} : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (cusp_form Γ k) :=
+{ zero := ⟨0, slash_action.mul_zero _, (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  by {intro a,
+      convert (zero_at_im_infty_submodule ℂ).zero_mem',
+      apply slash_action.mul_zero _ }⟩}
+
+instance has_nsmul : has_smul ℕ (cusp_form Γ k) :=
+⟨ λ c f,
+  { to_fun := (c : ℂ) • f,
```

Comment:
Can you just use `c • f` here, rather than `(c : ℂ) • f`?  Same on `zsmul` below.